src/app/static/src/app/store/projects/actions.ts

```diff
@@ -113,18 +113,25 @@ export const actions: ActionTree<ProjectsState, RootState> & ProjectsActions = {
         }
     },
 
-    async uploadVersionState(context) {
+    async queueVersionStateUpload(context) {
         const {state, commit} = context;
         if (state.currentVersion) {
-            if (!state.versionUploadPending) {
-                commit({type: ProjectsMutations.SetVersionUploadPending, payload: true});
-                setTimeout(() => {
+            // remove any existing queued upload, as this request should supersede it
+            commit({type: ProjectsMutations.ClearQueuedVersionUpload});
+
+            const queuedId: number = window.setInterval(() => {
+                // wait for any ongoing uploads to finish before starting a new one
+                if (!state.versionUploadInProgress) {
```

Comment:
Note using `setTimeout` here would mean that if a request is already in progress at this point, the requested upload never happens. Using `setInterval` ensures that the most recent state change is persisted, even if it has to wait a little while for a previous request to finish uploading.

Reply:
More or less, except that essentially the `setTimeout` we had in place was doing nothing other than delaying each request by 2 seconds! I think the original intention was that it would throttle requests, so if one was in flight, another would not be triggered... if that had been working however, it would have caused the first issue you reference - the missed uploads!

And yes, now there's always a queued request, but each incoming one supersedes the previous :+1: